core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
         this.types = source.getTypes();
     }
 
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseContext.parseFieldMatcher(parseFieldMatcher);
-    }
-
+    /**
+     * @return the parseFieldMatcher configured in the index settings
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
     }
 
     public void reset() {
```

Comment:
do you think it would be possible to get rid of this reset too and make sure that each operation that needs a reset actually uses a new instance of the context? Seems much cleaner.